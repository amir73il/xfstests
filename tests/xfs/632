#! /bin/bash
# SPDX-License-Identifier: GPL-2.0
# Copyright (c) 2019 Red Hat, Inc.  All Rights Reserved.
#
# FS QA Test No. 632
#
# Post-EOF preallocation defeat test with O_SYNC buffered I/O that repeatedly
# closes and reopens the files.
#

. ./common/preamble
_begin_fstest auto prealloc rw

. ./common/filter

_require_scratch
# _count_extents() needs the fiemap command
_require_xfs_io_command "fiemap"

_cleanup()
{
	# wait all background processes done
	wait
	cd /
	rm -r -f $tmp.*
}

_scratch_mkfs > "$seqres.full" 2>&1
_scratch_mount

# Write multiple files in parallel using synchronous buffered writes that
# repeatedly close and reopen the fails. Aim is to interleave allocations to
# fragment the files. Assuming we've fixed the synchronous write defeat, we can
# still trigger the same issue with a open/read/close on O_RDONLY files. We
# should not be triggering EOF preallocation removal on files we don't have
# permission to write, so until this is fixed it should fragment badly.  Typical
# problematic behaviour shows per-file extent counts of 50-350 whilst fixed
# behaviour typically demonstrates post-eof speculative delalloc growth in
# extent size (~6 extents for 50MB file).
#
# Failure is determined by golden output mismatch from _within_tolerance().

workfile=$SCRATCH_MNT/file
nfiles=32
wsize=4096
wcnt=1000

write_file()
{
	idx=$1

	$XFS_IO_PROG -f -s -c "pwrite -b 64k 0 50m" $workfile.$idx
}

read_file()
{
	idx=$1

	for ((cnt=0; cnt<wcnt; cnt++)); do
		$XFS_IO_PROG -f -r -c "pread 0 28" $workfile.$idx
	done
}

rm -f $workfile.*
for ((n=0; n<nfiles; n++)); do
	write_file $n > /dev/null 2>&1 &
	read_file $n > /dev/null 2>&1 &
done
wait

for ((n=0; n<nfiles; n++)); do
	count=$(_count_extents $workfile.$n)
	# Acceptible extent count range is 1-40
	_within_tolerance "file.$n extent count" $count 6 5 10 -v
done

status=0
exit
