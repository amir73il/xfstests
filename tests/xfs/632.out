QA output created by 632
file.0 extent count is in range
file.1 extent count is in range
file.2 extent count is in range
file.3 extent count is in range
file.4 extent count is in range
file.5 extent count is in range
file.6 extent count is in range
file.7 extent count is in range
file.8 extent count is in range
file.9 extent count is in range
file.10 extent count is in range
file.11 extent count is in range
file.12 extent count is in range
file.13 extent count is in range
file.14 extent count is in range
file.15 extent count is in range
file.16 extent count is in range
file.17 extent count is in range
file.18 extent count is in range
file.19 extent count is in range
file.20 extent count is in range
file.21 extent count is in range
file.22 extent count is in range
file.23 extent count is in range
file.24 extent count is in range
file.25 extent count is in range
file.26 extent count is in range
file.27 extent count is in range
file.28 extent count is in range
file.29 extent count is in range
file.30 extent count is in range
file.31 extent count is in range
